src/background/services/deduplicator.test.ts

```diff
@@ -0,0 +1,241 @@
+    shouldReject = false,
+  }: {
+    returnValue: T;
+    timeout?: number;
+    shouldReject?: boolean;
+  }) =>
+    jest.fn(
+      async (..._args: unknown[]) =>
+        new Promise((resolve, reject) => {
+          try {
+            if (shouldReject) {
+              reject(new Error('Test error'));
+            }
+            setTimeout(() => {
+              resolve(returnValue);
+            }, timeout);
+          } catch (e) {
+            reject(e);
+          }
+        }),
+    );
+
+  describe('Basic Deduplication', () => {
+    it('should call the original function only once for multiple simultaneous calls', async () => {
+      const returnValue = {
+        access_token: { value: 'value', access: { type: 'incoming' } },
+      };
+      const fn = createAsyncFn({ returnValue });
+      const dedupedFn = deduplicatorService.dedupe(fn);
+      const resultPromises = [dedupedFn(), dedupedFn(), dedupedFn()];
+      jest.runAllTimers();
+
+      const results = await Promise.all(resultPromises);
+      expect(results[0]).toBe(returnValue);
+      expect(results[1]).toBe(returnValue);
+      expect(results[2]).toBe(returnValue);
+      expect(fn).toHaveBeenCalledTimes(1);
+    });
+
+    it('should support different function signatures', async () => {
+      const fn1 = createAsyncFn({
+        returnValue: 'hello',
+        timeout: 100,
+      });
+      Object.defineProperty(fn1, 'name', { value: 'fn1' });
+      // createAsyncFn function returns an anonymous function, which is created using the new Promise constructor.
+      // it does not have a `name` property, so it does NOT have a key for deduplication service
+      const fn2 = createAsyncFn({
+        returnValue: 'world',
+        timeout: 400,
+      });
+      Object.defineProperty(fn2, 'name', { value: 'fn2' });
+      const dedupedFn1 = deduplicatorService.dedupe(fn1);
+      const dedupedFn2 = deduplicatorService.dedupe(fn2);
+      const resultPromises = [dedupedFn1('arg1'), dedupedFn2('arg2')];
+
+      jest.runAllTimers();
+
+      const [result1, result2] = await Promise.all(resultPromises);
+
+      expect(result1).toBe('hello');
+      expect(result2).toBe('world');
+      expect(fn1).toHaveBeenCalledTimes(1);
+      expect(fn2).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe('Cache Arguments Configuration', () => {
+    it('should differentiate calls with different arguments when cacheFnArgs is true', async () => {
+      const returnValue = 2n;
+      const fn = createAsyncFn({ returnValue });
+      const dedupedFn = deduplicatorService.dedupe(fn, {
+        cacheFnArgs: true,
+      });
+
+      dedupedFn('arg1');
+      dedupedFn('arg2');
+      jest.runAllTimers();
+
+      expect(fn).toHaveBeenCalledTimes(2);
+      expect(fn).toHaveBeenCalledWith('arg1');
+      expect(fn).toHaveBeenCalledWith('arg2');
+    });
+
+    it('should cache calls with the same arguments when cacheFnArgs is true', async () => {
+      const returnValue = 2n;
```

Comment:
Let's return an object (not a primitive), so we can know it's the same object (=== equal kind same) being returned as cached.